### PR TITLE
week2 homework

### PR DESCRIPTION
**Pull request by Enqira**

### CODE REVIEW

File: week2/task1/Program.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+var time = new JupiterTime();
+time.Hours = 8;
+time.Minutes = 5;
+
+PrintTime(time);
+
+void PrintTime(JupiterTime time)
+{
+    string f = "00.##";
+    Console.WriteLine($"{time.Hours.ToString(f)}:{time.Minutes.ToString(f)}");
```
Comment (ElhamDargi) [inline week2/task1/Program.cs:11]: Great!!
You can also write like this.

- Console.WriteLine($"{time.Hours:00}:{time.Minutes:00}");

File: week2/task2/Program.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+{
+    string f = "00.##";
+    Console.WriteLine($"{time.Hours.ToString(f)}:{time.Minutes.ToString(f)}");
+}
+public class JupiterTime
+{
+    int _hours = 0, _minutes = 0;
+
+    // Constructor
+    public JupiterTime(int hours = 0, int minutes = 0)
```
Comment (ElhamDargi) [inline week2/task2/Program.cs:18]: For constructor:
public JupiterTime( int hours, int minutes)    
    { 
        Hours = hours;
        Minutes = minutes;
    }

You set these conditions in the" set" and the "get" of the private fields later.

File: week2/task2/Program.cs [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    // Hours Prop
+    public int Hours
+    {
+        get
+        {
+            return _hours;
+        }
+        set
+        {
+            _hours += value;
```
Comment (ElhamDargi) [inline week2/task2/Program.cs:43]: You should handle "the negative and large number" here.
 if (value < 0) 
      _hours = (10 + _hours + value) % 10;
else if (value > 9)
      _hours = (value % 10);
else
      _hours = value;



It also works but for readable and clean code it's better to do it like this.

File: week2/task3/Program.cs [hunk 4]
```diff
@@ -0,0 +1,77 @@
+﻿// Task 3
+var time = new JupiterTime(2, 20);
+time.AddHours(-4);
```
Comment (ElhamDargi) [inline week2/task3/Program.cs:3]: "AddHours" has an int output and you should save it in a variable. It's better you consider the "JupiterTime" as an output. Then you can change it like this:

-  public JupiterTime AddHours(int count)

        {
            Hours = _hours + count;
            return new JupiterTime(_hours, _minutes);
        }

- var newTime = time.AddHours(-4);
   PrintTime(newTime);

File: week2/task4/Program.cs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+        set
+        {
+            //check if minutes is more than 59, and if so add the exeeding minutes to hours (eack hour is 60 minutes)
+            _minutes += value;
+            if (_minutes > 59)
+            {
+                _hours += _minutes / 60;
+                _minutes = _minutes % 60;
+            }
+
```
Comment (ElhamDargi) [inline week2/task4/Program.cs:63]: Well done!!!
For safety results, It's better to consider all conditions.

    set
            {
                if (value < 0)
                { 
                    throw new Exception("Invaid minutes");
                }
                else if (value > 59)
                {
                    _minutes = value % 60;
                    _hours = _hours + value / 60;
                    if (_hours > 10)
                    {
                        _hours = _hours % 10;
                    }
                }
                else
                    _minutes = value;
            }

File: week2/task6/Program.cs [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        string f = "00.##";
+        return $"{_hours.ToString(f)}:{_minutes.ToString(f)}";
+    }
+}
+
+
+// Signaler Class
+public class Signaler
+{
+    public List<JupiterTime> timers { get; set; } = new List<JupiterTime>();
```
Comment (ElhamDargi) [inline week2/task6/Program.cs:112]: Its not necessary to write {set; get;} here.

- public List<JupiterTime> timers = new List<JupiterTime>();

File: week2/task7/Program.cs [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        bool timerFound = false;
+        foreach (var item in timers)
+        {
+            if (time.Hours > item.Hours || (time.Hours == item.Hours && time.Minutes > item.Minutes))
+            {
+                Console.WriteLine(item.ToString());
+                timerFound = true;
+            }
+        }
+        if (!timerFound) Console.WriteLine("No signals needed to be sent yet.");
```
Comment (ElhamDargi) [inline week2/task7/Program.cs:146]: Excellent!!! 👍 
Very good solution.

File: week2/task3/Program.cs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    // Add hours method
+    public int AddHours(int hours)
+    {
+
+        _hours += hours;
+
+        // Hanlde negative hours, 
+        if (_hours < 0) _hours += System.Math.Abs(_hours - (_hours % 10) - 10);
+        if (_hours > 9) _hours = _hours % 10;
+
```
Comment (ElhamDargi) [inline week2/task3/Program.cs:74]: It is another way.

if (value < 0)
_hours = (10 + _hours + value) % 10;
else if (value > 9)
_hours = (value % 10);
else
_hours = value;

File: week2/task2/Program.cs (hunk 3 above)
Comment (Enqira) [reply week2/task2/Program.cs:43]: @ElhamDargi Yes, thank you, I did not know that I should have done that.

File: week2/task3/Program.cs (hunk 4 above)
Comment (Enqira) [reply week2/task3/Program.cs:3]: @ElhamDargi Yes, Thank you :)

**Pull request merged**